dali/kernels/signal/window/extract_windows_gpu.cuh

```diff
@@ -0,0 +1,568 @@
+    float w = window ? window[window_offset] : 1;
+    ptrdiff_t idx = window_start + window_offset;
+    if (reflect) {
+      idx = boundary::idx_reflect_101(idx, length);
+    }
+    float v = idx >= 0 && idx < length ? ConvertNorm<float>(src[idx]) * w : Src();
+    tmp[page][threadIdx.y][threadIdx.x] = v;
+  }
+  __syncthreads();
+
+  int win_ofs = blockIdx.y * kBlock + threadIdx.y;
+  ptrdiff_t win_idx = first_window_idx + threadIdx.x;
+
+  if (win_ofs < win_len && win_idx < num_windows)
+    dst[stride * win_ofs + win_idx] = ConvertSatNorm<Dst>(tmp[page][threadIdx.x][threadIdx.y]);
+}
+
+/**
+ * @brief Copy a sample to the windows containing it, applying the window function.
+ *
+ * This function takes one sample from the input and copies it to all windows that the sample
+ * contributes to.
+ * This function is optimized for cases when each sample contributes to the same number of windows.
+ *
+ * @param idx           Index of the sample processed by current thread, relative to `src`.
+ *                      If the index is out of range (<0 or >= length), the signal is reflected
+ *                      or padded with zeros.
+ * @param dst           destination buffer
+ * @param num_windows   maximum number of windows to extract from given input
+ * @param stride        stride, in elements, between consecutive windows in `dst`
+ * @param src           input signal
+ * @param length        length, in samples, of the signal
+ * @param window        window function
+ * @param win_len       length, in samples, of the window
+ * @param win_center    center of the window function; typically win_len/2
+ * @param step          step, in samples, between consecutive windows in the input
+ * @param reflect       if true, reflect the signal at ends, otherwise zero-pad
+ */
+template <typename Dst, typename Src>
+__device__ void ExtractHorizontalWindows(
+    ptrdiff_t idx,
+    Dst *__restrict__ dst, ptrdiff_t num_windows, ptrdiff_t stride,
+    const Src *__restrict__ src, ptrdiff_t length,
+    const float *__restrict__ window,
+    int win_len,
+    int win_center,
+    int step,
+    bool reflect) {
+
+  // calculate the index of the first window that sample at `idx ` contributes to
+  ptrdiff_t idx0 = idx + win_center - win_len + step;  // add step to round up
+  ptrdiff_t win0 = ::max(ptrdiff_t(), idx0/step);
+
+  float value = 0;
+
+  if (reflect) {
+    ptrdiff_t src_idx = boundary::idx_reflect_101(idx, length);
+    value = ConvertSatNorm<float>(src[src_idx]);
+  } else {
+    if (idx >= 0 && idx < length)
+      value = ConvertSatNorm<float>(src[idx]);
+  }
+
+  for (int win_ofs = idx - (win0 * step - win_center), win_idx = win0;
+      win_ofs >= 0 && win_idx < num_windows;
+      win_ofs -= step, win_idx++) {
+    float v = window ? value * window[win_ofs] : value;
+    dst[win_idx * stride + win_ofs] = ConvertSatNorm<Dst>(v);
+  }
+}
+
+template <typename Dst, typename Src>
+__global__ void ExtractVerticalWindowsKernel(
```

Comment:
What value `ExtractVerticalWindowsKernel` add to `ExtractVerticalWindowsBlock` invocation?
If only `blockIdx.x * kBlock` then maybe we should merge it together.

Reply:
You can't invoke a `__global__` function from another function - and `ExtractVerticalWindowsBlock` is used twice, with different parameters, in two different kernels.